### PR TITLE
Auto-add tracked repos to .gitignore

### PR DESCRIPTION
**Pull request by alibkord**

### CODE REVIEW

File: src/main/scala/org/stormenroute/mecha/package.scala [hunk 1]
```diff
@@ -432,4 +432,65 @@ package object mecha {
...
+      log.info(s"Adding '$toAdd' to '${ignoreFilePath.toString}'...")
+      writeFile(ignoreFilePath, add(readFile(ignoreFilePath), toAdd))
+    }
+
+    private def readFile(file: Path): Seq[String] = {
+      Files.readAllLines(file)
+    }
+
+    private def writeFile(file: Path, content: Seq[String]) = {
+      Files.write(file, content, WRITE, CREATE)
```
Comment (axel22) [inline src/main/scala/org/stormenroute/mecha/package.scala]: Seems like you need to call `content.toArray` here.

File: src/main/scala/org/stormenroute/mecha/package.scala [hunk 2]
```diff
@@ -432,4 +432,65 @@ package object mecha {
...
+
+    val log = MechaLog.Println
+
+    def ignore(ignoreFilePath: Path, toAdd: String) = {
+      log.info(s"Adding '$toAdd' to '${ignoreFilePath.toString}'...")
+      writeFile(ignoreFilePath, add(readFile(ignoreFilePath), toAdd))
+    }
+
+    private def readFile(file: Path): Seq[String] = {
+      Files.readAllLines(file)
```
Comment (axel22) [inline src/main/scala/org/stormenroute/mecha/package.scala]: You should specify a charset here, since our builders are still running on JDK 7.

**Comment by axel22** [comment]
Thank you very much for your contribution!

I would be happy to merge this, but there seem to be some build errors - could you ensure that the build passes?

**Comment by alibkord** [comment]
Thanks a lot for getting back to me. :-)
I'll fix the problems and get back to you asap!

**Comment by alibkord** [comment]
Looks like that fixed the builds.
When you're finished reviewing, I'll squash/rebase the commits into one.
Thanks again.

File: src/main/scala/org/stormenroute/mecha/package.scala [hunk 3]
```diff
@@ -432,4 +432,66 @@ package object mecha {
...
+    object Empty extends Line
+    case class Comment(c: String) extends Line
+    case class Pattern(p: String) extends Line {
+      val trimmed = p.trim
+      val negative = trimmed.startsWith("!")
+      lazy val matcher = FileSystems.getDefault.getPathMatcher(s"glob:./${cleanUp(trimmed)}")
+      override def matches(dirName: String): Boolean = {
+        val path = if (!dirName.startsWith("./")) s"./$dirName" else dirName
+        val matches = matcher.matches(Paths.get(path))
+        (!negative && matches) || (negative && !matches)
```
Comment (axel22) [inline src/main/scala/org/stormenroute/mecha/package.scala]: My understanding is that patterns starting with `!` in ignore/exclude files override blacklist, roughly speaking.

If so, won't this consider a line `alreadyIgnored` if there is a whitelist pattern that does not match the new line?
In other words, if you have this:

```
*.class
abc.class
```

And the new line that you want to insert is `foobar`, then `matches` returns `Some("!abc.class")` for `foobar`, even though `foobar` is not ignored.

You probably want to have two methods `blacklists` and `whitelists`, and conclude that something is already ignored if it `blacklists` returns non-`None`, but `whitelists` returns `None`. (this is assuming that I correctly understand that whitelisting takes precedence over blacklisting, in which case there's not much you can do after something is whitelisted)

Comment (alibkord) [inline src/main/scala/org/stormenroute/mecha/package.scala]: Excellent catch. I'll fix that.

Comment (axel22) [inline src/main/scala/org/stormenroute/mecha/package.scala]: Cool - thanks!

Btw, thinking a bit about the case where the folder that you're after is whitelisted, it seems to me that the best option is to report an error and disallow tracking until the problematic rules is fixed manually by the dev.

File: src/main/scala/org/stormenroute/mecha/package.scala [hunk 4]
```diff
@@ -432,4 +432,109 @@ package object mecha {
...
+    }
+
+    def ignore(toIgnore: String, gitIgnoreFilePath: Path, gitExcludeFilePath: Path) = {
+      val gitIgnore = if (Files.exists(gitIgnoreFilePath)) readFile(gitIgnoreFilePath) else Seq()
+      val gitExclude = if (Files.exists(gitExcludeFilePath)) readFile(gitExcludeFilePath) else Seq()
+
+      addIgnore(toIgnore, gitExclude, gitIgnore ++ gitExclude).fold(
+          {
+            case wl: Seq[WhitelistPattern] =>
+              log.warn(s"The new repo can not be ignored by Git. It is whitelisted by the following patterns: $wl")
```
Comment (alibkord) [inline src/main/scala/org/stormenroute/mecha/package.scala]: I think a WARN log should be enough so the repo is still cloned, but user can still see why it's not ignored and can manually fix it, but I'm happy to make it stricter and fail the process with an ERROR log if you prefer that.

File: src/main/scala/org/stormenroute/mecha/package.scala [hunk 5]
```diff
@@ -432,4 +432,109 @@ package object mecha {
...
+      Files.write(file, content, UTF_8, WRITE, CREATE)
+    }
+
+    def ignore(toIgnore: String, gitIgnoreFilePath: Path, gitExcludeFilePath: Path) = {
+      val gitIgnore = if (Files.exists(gitIgnoreFilePath)) readFile(gitIgnoreFilePath) else Seq()
+      val gitExclude = if (Files.exists(gitExcludeFilePath)) readFile(gitExcludeFilePath) else Seq()
+
+      addIgnore(toIgnore, gitExclude, gitIgnore ++ gitExclude).fold(
+          {
+            case wl: Seq[WhitelistPattern] =>
```
Comment (alibkord) [inline src/main/scala/org/stormenroute/mecha/package.scala]: I just realised this type will be erased in runtime! _facepalm_

File: src/main/scala/org/stormenroute/mecha/package.scala [hunk 6]
```diff
@@ -432,4 +432,113 @@ package object mecha {
...
+              log.info(s"The new repo is already ignored via the following pattern(s): $patterns")
+          },
+          b => writeFile(gitExcludeFilePath, b))
+    }
+
+    sealed trait Patterns
+    case class Whitelists(patterns: Seq[WhitelistPattern]) extends Patterns
+    case class Blacklists(patterns: Seq[BlacklistPattern]) extends Patterns
+
+    def addIgnore(toAdd: String, addTo: Seq[String], check: Seq[String]): Either[Patterns, Seq[String]] = {
```
Comment (axel22) [inline src/main/scala/org/stormenroute/mecha/package.scala]: Perhaps a less ambiguous naming could be: `(lineToAdd: String, excludeLines: Seq[String], allLines: Seq[String])`

**Comment by axel22** [comment]
LGTM, thanks a lot for this!

I will merge when you're ready rebasing/squashing (but this is not required - your choice, just let me know).

**Comment by alibkord** [comment]
Awesome thanks for the review. I'll feel better if I squash the commits so
I don't litter your history! ;-)
I don't know what timezone you're in, but hopefully it'll be ready in the
next 1.5-2 hours.

Cheers.

**Comment by axel22** [comment]
I'm in CEST (UTC+1). No rush - whenever you get around to it!

**Comment by alibkord** [comment]
~~Forgot one last thing: will edit the README to reflect this change.~~
Done!

**Comment by axel22** [comment]
Thanks for the awesome PR, this is indeed the thing that's been missing!

The PR should get merged automatically in a few hours.

**Comment by alibkord** [comment]
Awesome. Thank you for being so responsive. :-)

**Pull request merged**